src/state/include/com/amazonaws/kinesis/video/state/Include.h

```diff
@@ -73,6 +83,7 @@ struct __StateMachineState {
     UINT64 acceptStates;
     GetNextStateFunc getNextStateFn;
     ExecuteStateFunc executeStateFn;
+    ErrorHandlerStateFunc errorHandlerStateFunc;

```

Comment:
This API was internal API until we needed to expose it to common library and webrtc. Hence, we didn't version the structure. Need to be careful with the deployment of the changes.

Reply:
There should not be a problem with producer SDK since producer state machines (client and stream state machines) are defined in PIC. For WebRTC sdk, the signaling state machine is defined in WebRTC sdk itself. And this will be a breaking change. I'm planning to send a new PR for WebRTC sdk to update the signaling state machine.
Do we expect customers to use newer version of PIC and older version of WebRTC? Probably not. If they do, then there will be a problem.

Reply:
I don't believe any customers are actually using the state machine library as is so updating C producer and the WebRTC will sort it out. However, my main concern is whether you need a separate state transition hook per state? Perhaps, a better approach is to have a single callback function (the hook) specified in the state machine constructor with it's custom data and the first parameter being the state itself (the UINT64 value) that it's being transitioned to. This way, you won't need to change the state machine definitions in the implementation and only provide the new parameter to the create state machine API?

Reply:
@MushMal at a high level let me outline what we're trying to do here.  Today with the state machine there is get next function which determines what the next state is then there is an execute function.  A case which is not handled properly is when the transition from A --> B fails for some reason.  There are two obvious ways to go about it:

1. You can add an error handler function and attach it to each state (it won't be different for each state but it *could* be if needed), this will fire when the transition from A --> B failed for some reason and will react appropriately.  
2. You can introduce an actual error state (which wouldn't require any changes to the state machine impl) clients would simply add this new state, today we do not do this but we could.

We opted to go with (1) because today there is an established flow of states and we felt it would be more difficult to reason about those transitions if we added a new state.  If we were starting from scratch I would have strongly advocated for (2).  In any case, (1) is what we are trying to accomplish in this PR.

With this in mind I do not think this error handler function is a pre hook, it's specific to be invoked when the state transition from A --> B fails for some reason.  And even more specifically today the only thing we're adding here is backoff/jitter before we retry.  Today when you transition A --> A the step state machine function already adds delay. But if you go from A --> B --> A the retry count is reset and this can indeed go in a tight loop.  What we're hoping to do here is have a "global" exponential backoff that's shared across all streams / client.  

Also it is not a state transition function either.  If you're wanting to go from A-->B and `f` is your state transition function then once you execute `f` you are now at state B.  That's not what this is, basically what we're adding is if the transition from A --> B failed for any reason, now lets invoke the error handler.  I understand today we can never technically fail to get from A --> B, we will arrive at B regardless of whether the executefn failed or not (maybe not in all cases).

Having a single callback function in the state machine constructor would not allow for the possibility of different error handler functions for each state.  It's an interesting choice perhaps we can think more about it.

Reply:
Thanks for detailed explanation, Hassan! I don't have anything more to add to justify the design choice. @MushMal do you agree with this?
@hassanctech I like the idea of renaming it from `StateErrorHandlerFunc` to `StateTransitionHookFunc`. Transition hook name sounds more apt.